## imort library


In [2]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
import os

In [3]:
PROJ_DIR=r"/home/sunbeam/Documents/flight_price"
DATA_DIR="data"


In [4]:
def get_data(name):
    file_name=f"{name}.csv"
    file_path=os.path.join(PROJ_DIR,DATA_DIR,file_name)
    return pd.read_csv(file_path)
    

SyntaxError: unterminated string literal (detected at line 1) (3698387336.py, line 1)

In [5]:
flights=get_data("flight_price")


In [6]:
flights


Airline Date_of_Journey    Source Destination  \
0           IndiGo      24/03/2019  Banglore   New Delhi   
1        Air India       1/05/2019   Kolkata    Banglore   
2      Jet Airways       9/06/2019     Delhi      Cochin   
3           IndiGo      12/05/2019   Kolkata    Banglore   
4           IndiGo      01/03/2019  Banglore   New Delhi   
...            ...             ...       ...         ...   
10678     Air Asia       9/04/2019   Kolkata    Banglore   
10679    Air India      27/04/2019   Kolkata    Banglore   
10680  Jet Airways      27/04/2019  Banglore       Delhi   
10681      Vistara      01/03/2019  Banglore   New Delhi   
10682    Air India       9/05/2019     Delhi      Cochin   

                       Route Dep_Time  Arrival_Time Duration Total_Stops  \
0                  BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1      CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2      DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h     2 stops   
3            CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4            BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   
...                      ...      ...           ...      ...         ...   
10678              CCU → BLR    19:55         22:25   2h 30m    non-stop   
10679              CCU → BLR    20:45         23:20   2h 35m    non-stop   
10680              BLR → DEL    08:20         11:20       3h    non-stop   
10681              BLR → DEL    11:30         14:10   2h 40m    non-stop   
10682  DEL → GOI → BOM → COK    10:55         19:15   8h 20m     2 stops   

      Additional_Info  Price  
0             No info   3897  
1             No info   7662  
2             No info  13882  
3             No info   6218  
4             No info  13302  
...               ...    ...  
10678         No info   4107  
10679         No info   4145  
10680         No info   7229  
10681         No info  12648  
10682         No info  11753  

[10683 rows x 11 columns]

In [7]:
flights.sample(50
              )

Airline Date_of_Journey    Source Destination  \
4002              IndiGo       6/03/2019     Delhi      Cochin   
3708           Air India       1/05/2019    Mumbai   Hyderabad   
6745         Jet Airways      27/05/2019     Delhi      Cochin   
2819           Air India      24/06/2019     Delhi      Cochin   
5308         Jet Airways      06/03/2019  Banglore   New Delhi   
7448           Air India       9/03/2019     Delhi      Cochin   
3061              IndiGo      18/06/2019   Chennai     Kolkata   
6712         Jet Airways       6/03/2019     Delhi      Cochin   
2852           Air India      06/03/2019  Banglore   New Delhi   
3424   Multiple carriers       9/06/2019     Delhi      Cochin   
2749              IndiGo      15/05/2019     Delhi      Cochin   
9077         Jet Airways       6/06/2019  Banglore       Delhi   
2621         Jet Airways      12/03/2019  Banglore   New Delhi   
1256         Jet Airways      18/05/2019   Kolkata    Banglore   
459    Multiple carriers      15/06/2019     Delhi      Cochin   
1870           Air India       9/04/2019  Banglore       Delhi   
525            Air India       1/05/2019   Kolkata    Banglore   
849             Air Asia      27/05/2019     Delhi      Cochin   
5594              IndiGo       9/03/2019   Kolkata    Banglore   
7104           Air India      15/05/2019     Delhi      Cochin   
9316            SpiceJet      21/03/2019     Delhi      Cochin   
7599         Jet Airways      24/03/2019   Kolkata    Banglore   
3553         Jet Airways      21/03/2019     Delhi      Cochin   
10093             IndiGo      15/04/2019  Banglore       Delhi   
1034            SpiceJet      24/04/2019     Delhi      Cochin   
6952              IndiGo       9/06/2019   Kolkata    Banglore   
9090              IndiGo      24/06/2019   Kolkata    Banglore   
6665         Jet Airways      15/05/2019    Mumbai   Hyderabad   
42           Jet Airways      18/05/2019     Delhi      Cochin   
3830         Jet Airways      21/05/2019     Delhi      Cochin   
551          Jet Airways       3/03/2019     Delhi      Cochin   
10325           SpiceJet      21/03/2019  Banglore   New Delhi   
9351         Jet Airways      21/06/2019    Mumbai   Hyderabad   
4892              IndiGo       1/05/2019   Kolkata    Banglore   
5206            SpiceJet      15/04/2019   Kolkata    Banglore   
5132              IndiGo      12/04/2019  Banglore       Delhi   
4694           Air India       6/05/2019   Kolkata    Banglore   
6942              IndiGo      12/05/2019   Chennai     Kolkata   
2802         Jet Airways      06/03/2019  Banglore   New Delhi   
2769              IndiGo       9/04/2019     Delhi      Cochin   
10235             IndiGo      03/03/2019  Banglore   New Delhi   
9229         Jet Airways      12/06/2019  Banglore       Delhi   
9203         Jet Airways       1/06/2019  Banglore       Delhi   
4968              IndiGo       3/04/2019   Kolkata    Banglore   
8358              IndiGo      24/06/2019  Banglore       Delhi   
7154         Jet Airways       9/04/2019  Banglore       Delhi   
10602          Air India      18/05/2019     Delhi      Cochin   
5405         Jet Airways      24/05/2019   Kolkata    Banglore   
10110  Multiple carriers      18/05/2019     Delhi      Cochin   
4259           Air India       9/05/2019   Kolkata    Banglore   

                       Route Dep_Time  Arrival_Time Duration Total_Stops  \
4002         DEL → BBI → COK    02:35         08:30   5h 55m      1 stop   
3708               BOM → HYD    19:30         20:45   1h 15m    non-stop   
6745         DEL → BOM → COK    21:50  12:35 28 May  14h 45m      1 stop   
2819   DEL → HYD → MAA → COK    13:15  09:25 25 Jun  20h 10m     2 stops   
5308               BLR → DEL    19:55         22:35   2h 40m    non-stop   
7448         DEL → BLR → COK    06:10         23:00  16h 50m      1 stop   
3061               MAA → CCU    11:35         14:00   2h 25m    non-stop   
6712         DEL → BOM → COK    11:30     

In [8]:
flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


-  The data set contains 10,683 rows and 11 features
-   column "route and total stop " have missing values
-   The data type of some features are not appropriate




### Prelimnary Analysis


#### 1]check data Type


In [9]:
flights.dtypes

Airline            object
Date_of_Journey    object
Source             object
Destination        object
Route              object
Dep_Time           object
Arrival_Time       object
Duration           object
Total_Stops        object
Additional_Info    object
Price               int64
dtype: object

In [33]:
value=flights.Dep_Time.iloc[6]
value

'18:55'

#### 2]Check For Duplicates

In [11]:
flights.duplicated().sum()

220

In [12]:
(
    flights.loc[flights.duplicated(keep=False)].sort_values(["Airline","Date_of_Journey","Source","Destination"])
)

Airline Date_of_Journey    Source Destination                  Route  \
6321  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
9848  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
572   Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
8168  Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
1495  Air India       1/04/2019   Kolkata    Banglore  CCU → DEL → COK → BLR   
...         ...             ...       ...         ...                    ...   
2692   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2870   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
3711   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2634    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   
5470    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   

     Dep_Time  Arrival_Time Duration Total_Stops  \
6321    08:50  23:55 02 Mar   39h 5m     2 stops   
9848    08:50  23:55 02 Mar   39h 5m     2 stops   
572     21:10         23:55   2h 45m    non-stop   
8168    21:10         23:55   2h 45m    non-stop   
1495    10:00  01:20 02 Apr  15h 20m     2 stops   
...       ...           ...      ...         ...   
2692    05:45         08:35   2h 50m    non-stop   
2870    05:45         08:35   2h 50m    non-stop   
3711    20:30         23:20   2h 50m    non-stop   
2634    11:30         14:10   2h 40m    non-stop   
5470    11:30         14:10   2h 40m    non-stop   

                   Additional_Info  Price  
6321                       No info  17135  
9848                       No info  17135  
572                        No info   7591  
8168                       No info   7591  
1495                       No info  10408  
...                            ...    ...  
2692  No check-in baggage included   4273  
2870  No check-in baggage included   4273  
3711  No check-in baggage included   3873  
2634                       No info   5403  
5470                       No info   5403  

[415 rows x 11 columns]

#### oberavations from premlinary Analyis

 -
  The data type of "Date of journy" ,"Dep_time "and "Arival_time" should be  change to Dtaetime
 - The type of "Duration" and "Stops " is mixed .its should be numeric type
 - There are 220  duplicates .These Should be Removed
   

## 4 Detail Analysis

#### Airline

In [37]:
flights.Airline

0             IndiGo
1          Air India
2        Jet Airways
3             IndiGo
4             IndiGo
            ...     
10678       Air Asia
10679      Air India
10680    Jet Airways
10681        Vistara
10682      Air India
Name: Airline, Length: 10683, dtype: object

In [38]:
flights.Airline.unique()

array(['IndiGo', 'Air India', 'Jet Airways', 'SpiceJet',
       'Multiple carriers', 'GoAir', 'Vistara', 'Air Asia',
       'Vistara Premium economy', 'Jet Airways Business',
       'Multiple carriers Premium economy', 'Trujet'], dtype=object)

In [43]:
(
    flights.Airline.str.replace("Premium economy","").str.replace("Business","")
    .str.title()
    .unique()
)

array(['Indigo', 'Air India', 'Jet Airways', 'Spicejet',
       'Multiple Carriers', 'Goair', 'Vistara', 'Air Asia', 'Vistara ',
       'Jet Airways ', 'Multiple Carriers ', 'Trujet'], dtype=object)

 - some of the entries have inconsistent/incorrect values
 -  Update the 'Airline' column in the existing DataFrame
  flights['Airline'] = flights['Airline'].str.replace("Premium economy", "")
                                    .replace("Business", "")
                                    .str.title()

 - Optionally, you can also get the unique values if needed
  unique_airlines = flights['Airline'].unique()


#### Date_of_journey

In [57]:
flights.Date_of_Journey

0        24/03/2019
1         1/05/2019
2         9/06/2019
3        12/05/2019
4        01/03/2019
            ...    
10678     9/04/2019
10679    27/04/2019
10680    27/04/2019
10681    01/03/2019
10682     9/05/2019
Name: Date_of_Journey, Length: 10683, dtype: object

In [58]:
pd.to_datetime(flights.Date_of_Journey,dayfirst=True)

0       2019-03-24
1       2019-05-01
2       2019-06-09
3       2019-05-12
4       2019-03-01
           ...    
10678   2019-04-09
10679   2019-04-27
10680   2019-04-27
10681   2019-03-01
10682   2019-05-09
Name: Date_of_Journey, Length: 10683, dtype: datetime64[ns]

#### Source

In [66]:
flights.Source.unique()

array(['Banglore', 'Kolkata', 'Delhi', 'Chennai', 'Mumbai'], dtype=object)

#### Destination

In [69]:
flights.Destination.unique()

array(['New Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Delhi', 'Hyderabad'],
      dtype=object)

#### Route

 - same information is available from number of stops and desination ,source
 - because of this reason we are going to drop this feature

#### Dep_Time

In [77]:
flights.Dep_Time.dtype

dtype('O')

In [78]:
(
    flights
    .Dep_Time
    .loc[lambda ser:ser.str.contains("[^0-9:]")]

    
)

Series([], Name: Dep_Time, dtype: object)

 - Here we validate that Dep_time does not conatins any specail numeric symbol than ":"


In [80]:
pd.to_datetime(flights.Dep_Time).dt.time

/tmp/ipykernel_53087/1056054517.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  pd.to_datetime(flights.Dep_Time).dt.time


0        22:20:00
1        05:50:00
2        09:25:00
3        18:05:00
4        16:50:00
           ...   
10678    19:55:00
10679    20:45:00
10680    08:20:00
10681    11:30:00
10682    10:55:00
Name: Dep_Time, Length: 10683, dtype: object

#### Arrival_Time

In [83]:
flights.Arrival_Time

0        01:10 22 Mar
1               13:15
2        04:25 10 Jun
3               23:30
4               21:35
             ...     
10678           22:25
10679           23:20
10680           11:20
10681           14:10
10682           19:15
Name: Arrival_Time, Length: 10683, dtype: object

 - From above we can see that this feature conatins mixed value like date and time
 - we have to take only the time part


In [91]:
(
    flights
    .Arrival_Time
    .loc[lambda ser:ser.str.contains("[^0-9:]")]
    .str.split(" ",n=1)
    .str.get(-1)
    .unique()
    
)

array(['22 Mar', '10 Jun', '13 Mar', '02 Mar', '10 May', '04 Mar',
       '13 Jun', '28 May', '19 Mar', '07 May', '02 Jun', '16 Jun',
       '19 May', '16 May', '28 Jun', '02 May', '28 Mar', '19 Jun',
       '04 Apr', '25 Mar', '07 Mar', '25 Jun', '07 Jun', '25 May',
       '13 May', '16 Mar', '22 May', '10 Apr', '04 Jun', '20 May',
       '28 Apr', '25 Apr', '10 Mar', '19 Apr', '13 Apr', '02 Apr',
       '23 Mar', '22 Apr', '11 May', '07 Apr', '03 May', '08 Mar',
       '03 Mar', '05 Mar', '22 Jun', '04 May', '26 May', '16 Apr',
       '26 Jun', '29 May', '29 Jun', '29 Mar', '23 May', '17 Jun'],
      dtype=object)

#### Duration

In [95]:
flights.Duration

0        2h 50m
1        7h 25m
2           19h
3        5h 25m
4        4h 45m
          ...  
10678    2h 30m
10679    2h 35m
10680        3h
10681    2h 40m
10682    8h 20m
Name: Duration, Length: 10683, dtype: object

In [129]:
(
    flights.Duration
    .drop(index=6474)
    .str.split(" ",expand=True)
    .set_axis(["hours","minute"],axis=1)
    .assign(
        hours=lambda df_:
        (
           df_
            .hours
            .str.replace("h","")
            .astype(int)
            .mul(60)
        ),
        minute=lambda df_:
        (
            df_
            .minute
            .str.replace("m","")
            .fillna("0")
            .astype(int)
        )
        
    )
    .sum(axis=1)
    
    
    
   
     

    
   
    
)

0         170
1         445
2        1140
3         325
4         285
         ... 
10678     150
10679     155
10680     180
10681     160
10682     500
Length: 10682, dtype: int64

In [133]:
(
    flights.Duration
    .drop(index=6474)
    .str.split(" ",expand=True)
    .set_axis(["hours","minute"],axis=1)
    .assign(
        hours=lambda df_:
        (
           df_
            .hours
            .str.replace("h","")
            .astype(int)
            .mul(60)
        ),
        minute=lambda df_:
        (
            df_
            .minute
            .str.replace("m","")
            .fillna("0")
            .astype(int)
        )
        
    )
    .sum(axis=1)
    .rename("duration_minutes")
    .to_frame()
    .join(flights.Duration.drop(index=6474))
        
    
    
   
     

    
   
    
)

duration_minutes Duration
0                   170   2h 50m
1                   445   7h 25m
2                  1140      19h
3                   325   5h 25m
4                   285   4h 45m
...                 ...      ...
10678               150   2h 30m
10679               155   2h 35m
10680               180       3h
10681               160   2h 40m
10682               500   8h 20m

[10682 rows x 2 columns]

In [109]:
(
    flights
    .Duration
    .loc[lambda ser: ~ser.str.contains("m")]
    .unique()
)

array(['19h', '23h', '22h', '12h', '3h', '5h', '10h', '18h', '24h', '15h',
       '16h', '8h', '14h', '20h', '13h', '11h', '9h', '27h', '26h', '4h',
       '7h', '30h', '21h', '28h', '47h', '6h', '25h', '38h', '34h'],
      dtype=object)

In [111]:
(
    flights
    .Duration
    .loc[lambda ser: ~ser.str.contains("h")]
    
)

6474    5m
Name: Duration, dtype: object

In [112]:
flights.iloc[[6474]]

Airline Date_of_Journey  Source Destination                  Route  \
6474  Air India       6/03/2019  Mumbai   Hyderabad  BOM → GOI → PNQ → HYD   

     Dep_Time Arrival_Time Duration Total_Stops Additional_Info  Price  
6474    16:50        16:55       5m     2 stops         No info  17327

- The Observation index 6474 has duration of 5
- This is clearly wrong
- will delete this observation

In [29]:
flights.select_dtypes(include="O").columns

Index(['Airline', 'Date_of_Journey', 'Source', 'Destination', 'Route',
       'Dep_Time', 'Arrival_Time', 'Duration', 'Total_Stops',
       'Additional_Info'],
      dtype='object')

-
  i want to remove the extra spaces from strip  how will i will i do

-  flights["airline"]=flights["airline"].str.strip()
-  its overriding  the current dateframe
-  df_ will hold the cureent state of the object

## 5 Cleaning Operations

In [142]:
def convert_to_minutes(ser):
        return 
            (
                ser
               
        
        .str.split(" ",expand=True)
        .set_axis(["hours","minute"],axis=1)
        .assign(
            hours=lambda df_:
            (
               df_
                .hours
                .str.replace("h","")
                .astype(int)
                .mul(60)
            ),
            minute=lambda df_:
            (
                df_
                .minute
                .str.replace("m","")
                .fillna("0")
                .astype(int)
            )
            
        )
        .sum(axis=1)
        
                
                
        
        
                
                
                
            )

IndentationError: unexpected indent (884756643.py, line 3)

In [140]:
def clean_data(df):
    return (df

            .drop(index=[6474])

            .assign(**{
                col:df[col].str.strip()
                for col in df.select_dtypes(include="O").columns
                
            })

            .rename(columns=str.lower)

            .assign(
                airline=lambda df_:
                (
                    df_
                    .airline
                    .str.replace("Premium economy","")
                    .str.replace("Business","")
                    .str.strip()
                    .str.title()
    
                    
                    
                ),
                date_of_journey=lambda df_:
                (
                    pd.to_datetime(df_.date_of_journey,dayfirst=True)
                ),
                dep_time=lambda df_:
                (
                   pd.to_datetime(df_.dep_time).dt.time 
                ),
                arrival_time=lambda df_:
                (
                   pd.to_datetime(df_.arrival_time).dt.time 
                ),
                duration=lambda df_:df_.duration.pipe(convert_to_minutes)
                
                
                
                
                
            )
            .drop(columns="route")




            
            
                
            )
            
           
           
           
           
    
    
    

In [141]:
clean_data(flights)

/tmp/ipykernel_53087/3350242145.py:33: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  pd.to_datetime(df_.dep_time).dt.time
/tmp/ipykernel_53087/3350242145.py:37: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  pd.to_datetime(df_.arrival_time).dt.time


airline date_of_journey    source destination  dep_time  \
0           Indigo      2019-03-24  Banglore   New Delhi  22:20:00   
1        Air India      2019-05-01   Kolkata    Banglore  05:50:00   
2      Jet Airways      2019-06-09     Delhi      Cochin  09:25:00   
3           Indigo      2019-05-12   Kolkata    Banglore  18:05:00   
4           Indigo      2019-03-01  Banglore   New Delhi  16:50:00   
...            ...             ...       ...         ...       ...   
10678     Air Asia      2019-04-09   Kolkata    Banglore  19:55:00   
10679    Air India      2019-04-27   Kolkata    Banglore  20:45:00   
10680  Jet Airways      2019-04-27  Banglore       Delhi  08:20:00   
10681      Vistara      2019-03-01  Banglore   New Delhi  11:30:00   
10682    Air India      2019-05-09     Delhi      Cochin  10:55:00   

      arrival_time duration total_stops additional_info  price  
0         01:10:00     None    non-stop         No info   3897  
1         13:15:00     None     2 stops         No info   7662  
2         04:25:00     None     2 stops         No info  13882  
3         23:30:00     None      1 stop         No info   6218  
4         21:35:00     None      1 stop         No info  13302  
...            ...      ...         ...             ...    ...  
10678     22:25:00     None    non-stop         No info   4107  
10679     23:20:00     None    non-stop         No info   4145  
10680     11:20:00     None    non-stop         No info   7229  
10681     14:10:00     None    non-stop         No info  12648  
10682     19:15:00     None     2 stops         No info  11753  

[10682 rows x 10 columns]

In [61]:
flights.dtypes

Airline            object
Date_of_Journey    object
Source             object
Destination        object
Route              object
Dep_Time           object
Arrival_Time       object
Duration           object
Total_Stops        object
Additional_Info    object
Price               int64
dtype: object

## 6 Split The Data

## 7 Export the Split data